webapp/src/tutorialValidator.ts

```diff
@@ -0,0 +1,112 @@
+/// <reference path="../../built/pxtlib.d.ts" />
+
+import * as compiler from "./compiler";
+
+export enum TutorialCodeStatus {
+    Unknown = "Unknown",
+    Invalid = "Invalid",
+    Valid = "Valid"
+}
+
+/**
+* Check the user's code to the tutorial and returns a Tutorial status of the user's code
+* @param step the current tutorial step
+* @param source XML of workspace
+* @param workspaceBlocks Blockly blocks used of workspace
+* @param blockinfo Typescripts of the workspace
+* @return A TutorialCodeStatus
+*/
+export async function validate(step: pxt.tutorial.TutorialStepInfo, source: string, workspaceBlocks: Blockly.Block[], blockinfo: pxtc.BlocksInfo): Promise<TutorialCodeStatus> {
+    // Check to make sure blocks are in the workspace
+    if (workspaceBlocks.length > 0) {
+        let userBlockTypes: string[] = workspaceBlocks.map(b => b.type);
+        let tutorialBlockTypes: string[] = await tutorialBlockList(step, blockinfo);
+        let usersBlockUsed: pxt.Map<number> = blockCount(userBlockTypes);
+        let tutorialBlockUsed: pxt.Map<number> = blockCount(tutorialBlockTypes);
+        if (!validateNumberOfBlocks(usersBlockUsed, tutorialBlockUsed)) {
+            return TutorialCodeStatus.Invalid;
+        }
+        return TutorialCodeStatus.Valid;
+    }
+    return TutorialCodeStatus.Unknown;
+}
+
+/**
+* Loops through an array of blocks and returns a map of blocks and the count for that block
+* @param arr a string array of blocks
+* @return a map <Block type, frequency>
+*/
+function blockCount(arr: string[]): pxt.Map<number> {
+    let frequencyMap: pxt.Map<number> = {};
+    for (let i: number = 0; i < arr.length; i++) {
+        let blockIndex = hasBlock(arr, i, arr[i]);
+        if (blockIndex == i) {
+            frequencyMap[arr[i]] = 0;
+        }
+        frequencyMap[arr[i]] = frequencyMap[arr[i]] + 1;
+    }
+    return frequencyMap;
+}
+
+/**
+* Checks to see if the array has the target value and returns the target value's index
```

Comment:
would the the indexOf function work for this? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf